// @vue/devtools-api@8.0.3/dist/index.cjs
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let __vue_devtools_kit = require("@vue/devtools-kit");
__vue_devtools_kit = __toESM(__vue_devtools_kit);

Object.defineProperty(exports, 'addCustomCommand', {
  enumerable: true,
  get: function () {
    return __vue_devtools_kit.addCustomCommand;
  }
});
Object.defineProperty(exports, 'addCustomTab', {
  enumerable: true,
  get: function () {
    return __vue_devtools_kit.addCustomTab;
  }
});
Object.defineProperty(exports, 'onDevToolsClientConnected', {
  enumerable: true,
  get: function () {
    return __vue_devtools_kit.onDevToolsClientConnected;
  }
});
Object.defineProperty(exports, 'onDevToolsConnected', {
  enumerable: true,
  get: function () {
    return __vue_devtools_kit.onDevToolsConnected;
  }
});
Object.defineProperty(exports, 'removeCustomCommand', {
  enumerable: true,
  get: function () {
    return __vue_devtools_kit.removeCustomCommand;
  }
});
Object.defineProperty(exports, 'setupDevToolsPlugin', {
  enumerable: true,
  get: function () {
    return __vue_devtools_kit.setupDevToolsPlugin;
  }
});
Object.defineProperty(exports, 'setupDevtoolsPlugin', {
  enumerable: true,
  get: function () {
    return __vue_devtools_kit.setupDevToolsPlugin;
  }
});
